UPGRADING.md

```diff
@@ -0,0 +1,207 @@
+
+Most state components, and the `State` object itself, now request at least some
+arguments to be passed by keyword. In most cases, it's likely that you were
+already doing this, but the API is now enforced.
+
+```python
+# Older Scenario code.
+container1 = Container('foo', True)
+state = State({'key': 'value'}, [relation1, relation2], [network], [container1, container2])
+
+# Scenario 7.x
+container1 = Container('foo', can_connect=True)
+state = State(
+    config={'key': 'value'},
+    relations={relation1, relation2},
+    networks={network},
+    containers={container1, container2},
+)
+```
+
+### Copying objects
+
+The `copy()` and `replace()` methods of `State` and the various state components
+have been removed. You should use the `dataclasses.replace` and `copy.deepcopy`
+methods instead.
+
+```python
+# Older Scenario code.
+new_container = container.replace(can_connect=True)
+duplicate_relation = relation.copy()
+
+# Scenario 7.x
+new_container = dataclasses.replace(container, can_connect=True)
+duplicate_relation = copy.deepcopy(relation)
+```
+
+### Mocking Container.exec
+
+In addition to the new ability to assert against content the charm writes to
+`stdin` and to use pattern matching for the command, `Container.exec_mocks` is
+now named `Container.execs`, `Container.service_status` is now named
+`Container.service_statuses`, and `ExecOutput` is now named `Exec`.
+
+### Resources are classes, not dictionaries
+
+The resources in State objects are now `scenario.Resource` objects rather than
+plain dictionaries, aligning with all of the other State components.
+
+```python
+# Older Scenario code
+state = State(resources={"/path/to/foo", pathlib.Path("/mock/foo")})
+
+# Scenario 7.x
+resource = Resource(location="/path/to/foo", source=pathlib.Path("/mock/foo"))
+state = State(resources={resource})
+```
+
+### State.storage and State.stored_state plurals
+
+The `State.storage` and `State.stored_state` attributes are now plurals, to
+reflect that you may have more than one in the state, aligning with the other
+State components.
+
+```python
+# Older Scenario code
+state = State(stored_state=[ss1, ss2], storage=[s1, s2])
+
+# Scenario 7.x
+state = State(stored_states={s1, s2}, storages={s1, s2})
+```
+
+### Network
+
+`Network` objects are added to the state as a set, like the other components,
+rather than as a dictionary of `{binding_name: network}` as previously. This
+means that the `Network` object now requires a binding name to be passed in when
+it is created.
+
+```python
+# Older Scenario code
+state = State(networks={"foo": Network.default()})
+
+# Scenario 7.x
+state = State(networks={Network.default("foo")})
+```
+
+### Names made private
+
+Several attributes and classes that were never intended for end users have been made private:
+
+* The `data_type_name` attribute of `StoredState` is now private.
+* The `RelationBase` class is now private.
+* The `Event` class is now private.
+
+### Minor removed functionality
```

Comment:
Yeah, it feels like none of these deserve a full section, since they are all basically "stop using the previously deprecated thing X - you could do this other thing instead". On the other hand, maybe a bunch of small sections is actually fine?

For now, I've adjusted the title to align with the other ones, so maybe it's a bit better?

Reply:
Small sections with the main point in the title would make these release notes very scannable, imo.